File: swift/Sources/Models/Categoriae/Goals.swift
// Goals.swift
// Domain entity representing objectives and targets
//
// Written by Claude Code on 2025-10-18
// Ported from Python implementation (python/categoriae/goals.py)
//
// This uses class-based inheritance to demonstrate Swift's object-oriented features
// Hierarchy: Goal → SmartGoal
//            Goal → Milestone

import Foundation

/// Base class for all goal types
///
/// Goals represent objectives with optional measurements and time bounds.
/// This is the most flexible form - all fields are optional.
///
/// Conforms to Persistable: id, friendlyName, detailedDescription, freeformNotes, logTime
class Goal: Persistable {
    // MARK: - Persistable Protocol Requirements

    var id: UUID
    var friendlyName: String?
    var detailedDescription: String?
    var freeformNotes: String?
    var logTime: Date

    // MARK: - Goal-specific Properties

    /// Type identifier for polymorphic storage ("Goal", "SmartGoal", "Milestone")
    var goalType: String { return "Goal" }

    /// What unit to measure (e.g., "km", "hours", "pages")
    var measurementUnit: String?

    /// Target value to achieve
    var measurementTarget: Double?

    /// When the goal period starts
    var startDate: Date?

    /// When the goal should be achieved by
    var targetDate: Date?

    /// Why this goal matters (SMART: Relevant)
    var howGoalIsRelevant: String?

    /// How to achieve this goal (SMART: Actionable)
    var howGoalIsActionable: String?

    /// Expected duration in weeks (e.g., 10 for ten-week term)
    var expectedTermLength: Int?

    // MARK: - Initialization

    /// Create a new goal with required and optional fields
    /// - Parameters:
    ///   - id: Unique identifier (auto-generated if not provided)
    ///   - friendlyName: Short description of the goal
    ///   - detailedDescription: Optional elaboration
    ///   - freeformNotes: Optional freeform notes
    ///   - logTime: When goal was created (defaults to now)
    ///   - measurementUnit: Optional unit (e.g., "km")
    ///   - measurementTarget: Optional target value
    ///   - startDate: Optional goal start date
    ///   - targetDate: Optional goal completion date
    ///   - howGoalIsRelevant: Optional relevance statement
    ///   - howGoalIsActionable: Optional action plan
    ///   - expectedTermLength: Optional duration in weeks
    init(
        id: UUID = UUID(),
        friendlyName: String? = nil,
        detailedDescription: String? = nil,
        freeformNotes: String? = nil,
        logTime: Date = Date(),
        measurementUnit: String? = nil,
        measurementTarget: Double? = nil,
        startDate: Date? = nil,
        targetDate: Date? = nil,
        howGoalIsRelevant: String? = nil,
        howGoalIsActionable: String? = nil,
        expectedTermLength: Int? = nil
    ) {
        self.id = id
        self.friendlyName = friendlyName
        self.detailedDescription = detailedDescription
        self.freeformNotes = freeformNotes
        self.logTime = logTime
        self.measurementUnit = measurementUnit
        self.measurementTarget = measurementTarget
        self.startDate = startDate
        self.targetDate = targetDate
        self.howGoalIsRelevant = howGoalIsRelevant
        self.howGoalIsActionable = howGoalIsActionable
        self.expectedTermLength = expectedTermLength
    }

    // MARK: - Validation Methods

    /// Check if this goal has defined start and target dates
    func isTimeBound() -> Bool {
        return startDate != nil && targetDate != nil
    }

    /// Check if this goal has a measurement unit and target
    func isMeasurable() -> Bool {
        return measurementUnit != nil && measurementTarget != nil
    }

    /// General validation - can be overridden by subclasses
    func isValid() -> Bool {
        // Base Goal has no strict requirements
        // Just check that if we have a target, it's positive
        if let target = measurementTarget, target <= 0 {
            return false
        }

        // Check date ordering if both dates exist
        if let start = startDate, let end = targetDate, start >= end {
            return false
        }

        return true
    }
}

// MARK: - SmartGoal Subclass

/// SMART goal with strict validation and required fields
///
/// Enforces all SMART criteria:
/// - Specific: Clear description (via friendlyName)
/// - Measurable: Has unit and target value (required)
/// - Achievable: Target is positive, has action plan (required)
/// - Relevant: Has relevance statement (required)
/// - Time-bound: Has start and end dates (required)
///
/// Example: "Run 120km in 10 weeks starting Oct 10"
class SmartGoal: Goal {
    // MARK: - Override goalType

    override var goalType: String { return "SmartGoal" }

    // MARK: - Initialization with Validation

    /// Create a SMART goal with all required fields
    ///
    /// All SMART criteria fields are required parameters (non-optional in init)
    /// Throws fatalError if validation fails (design choice - could use throws instead)
    init(
        id: UUID = UUID(),
        friendlyName: String? = nil,
        detailedDescription: String? = nil,
        freeformNotes: String? = nil,
        logTime: Date = Date(),
        measurementUnit: String,          // Required for SMART
        measurementTarget: Double,        // Required for SMART
        startDate: Date,                  // Required for SMART
        targetDate: Date,                 // Required for SMART
        howGoalIsRelevant: String,        // Required for SMART
        howGoalIsActionable: String,      // Required for SMART
        expectedTermLength: Int? = nil
    ) {
        // Call parent initializer with required fields
        super.init(
            id: id,
            friendlyName: friendlyName,
            detailedDescription: detailedDescription,
            freeformNotes: freeformNotes,
            logTime: logTime,
            measurementUnit: measurementUnit,
            measurementTarget: measurementTarget,
            startDate: startDate,
            targetDate: targetDate,
            howGoalIsRelevant: howGoalIsRelevant,
            howGoalIsActionable: howGoalIsActionable,
            expectedTermLength: expectedTermLength
        )

        // Validate SMART criteria after initialization
        validateSmartCriteria()
    }

    // MARK: - Validation

    private func validateSmartCriteria() {
        // Measurable: must have unit and positive target
        guard let unit = measurementUnit, !unit.isEmpty else {
            fatalError("SmartGoal requires measurementUnit (Measurable)")
        }

        guard let target = measurementTarget, target > 0 else {
            fatalError("SmartGoal requires positive measurementTarget (Measurable, Achievable)")
        }

        // Time-bound: must have valid date range
        guard let start = startDate, let end = targetDate else {
            fatalError("SmartGoal requires both startDate and targetDate (Time-bound)")
        }

        guard start < end else {
            fatalError("SmartGoal startDate must be before targetDate (Time-bound)")
        }

        // Relevant: must have relevance statement
        guard let relevant = howGoalIsRelevant, !relevant.isEmpty else {
            fatalError("SmartGoal requires howGoalIsRelevant (Relevant)")
        }

        // Achievable: must have action plan
        guard let actionable = howGoalIsActionable, !actionable.isEmpty else {
            fatalError("SmartGoal requires howGoalIsActionable (Achievable)")
        }
    }

    override func isValid() -> Bool {
        // SmartGoal is valid if it passes all SMART criteria
        // Since we validate in init, a SmartGoal is always valid if it exists
        return super.isValid() // Also check parent's basic rules
    }
}

// MARK: - Milestone Subclass

/// A significant checkpoint within a larger goal
///
/// Milestones are point-in-time targets, not ranges.
/// They require a target date but not a start date.
///
/// Example: "Reach 50km by week 5"
class Milestone: Goal {
    // MARK: - Override goalType

    override var goalType: String { return "Milestone" }

    // MARK: - Initialization with Validation

    /// Create a milestone with required target date
    ///
    /// Milestones don't have start dates (they're point-in-time)
    init(
        id: UUID = UUID(),
        friendlyName: String? = nil,
        detailedDescription: String? = nil,
        freeformNotes: String? = nil,
        logTime: Date = Date(),
        measurementUnit: String? = nil,
        measurementTarget: Double? = nil,
        targetDate: Date,                 // Required for Milestone
        howGoalIsRelevant: String? = nil,
        howGoalIsActionable: String? = nil,
        expectedTermLength: Int? = nil
    ) {
        // Call parent initializer
        super.init(
            id: id,
            friendlyName: friendlyName,
            detailedDescription: detailedDescription,
            freeformNotes: freeformNotes,
            logTime: logTime,
            measurementUnit: measurementUnit,
            measurementTarget: measurementTarget,
            startDate: nil,  // Milestones don't have start dates
            targetDate: targetDate,
            howGoalIsRelevant: howGoalIsRelevant,
            howGoalIsActionable: howGoalIsActionable,
            expectedTermLength: expectedTermLength
        )
    }

    override func isValid() -> Bool {
        // Milestone must have a target date
        guard targetDate != nil else {
            return false
        }

        // And must NOT have a start date (point-in-time, not range)
        guard startDate == nil else {
            return false
        }

        return super.isValid()
    }
}

File: python - inactive/categoriae/goals.py
"""
Goals are things we try to achieve within a window of time

This file contains conceptual sketches to understand entity relationships.
Written by Claude Code on 2025-10-09
Refactored to use dataclasses on 2025-10-16


"""

from abc import ABC
from dataclasses import dataclass, field
from datetime import datetime, timedelta
from typing import Optional

from categoriae.ontology import PersistableEntity


@dataclass
class ThingIWant(PersistableEntity):
    """
    Parent class: Broadest concept - things that have a description.

    This is the required essence: all things I might want must be describable.
    """
    pass


@dataclass(unsafe_hash=True)
class Goal(ThingIWant):
    """
    A general objective that may or may not be time-bound.

    Goals can be aspirational ("run a marathon someday") or have
    loose targets. They're more flexible than SmartGoals.

    Field structure:
    - Required: description (inherited from ThingIWant)
    - Persistence: id, created_at (like PersistableEntity pattern)
    - Optional: All measurement and time-bound fields

    Note: unsafe_hash=True allows Goal objects to be used as dictionary keys
    (required by progress_aggregation.py) while remaining mutable.
    """
    # Persistence infrastructure (with defaults, like PersistableEntity)
    goal_type: str = 'Goal'  # Class identifier for polymorphic storage

    # Optional goal-specific fields
    measurement_unit: Optional[str] = None
    measurement_target: Optional[float] = 0.0
    start_date: Optional[datetime] = None
    target_date: Optional[datetime] = None  # When goal should be achieved
    how_goal_is_relevant: Optional[str] = None
    how_goal_is_actionable: Optional[str] = None
    expected_term_length: Optional[int] = None  # e.g., 10 for 10-week term



    def is_time_bound(self) -> bool:
        """Check if this goal has defined start and target dates"""
        return self.start_date is not None and self.target_date is not None

    def is_measurable(self) -> bool:
        """Check if this goal has a measurement unit and target"""
        return self.measurement_unit is not None and self.measurement_target is not None


@dataclass
class Milestone(Goal):
    """
    A significant checkpoint within a larger goal or term.

    Milestones are concrete, dated targets that mark progress toward bigger goals.
    Unlike full SmartGoals, they're more like waypoints.

    Examples:
    - "Reach 50km by week 5" (milestone toward 120km goal)
    - "Complete chapter 3 by Nov 15" (milestone toward book completion)

    This is a child of Goal: ThingIWant → Goal → Milestone
    """
    goal_type: str = 'Milestone'  # Override parent's goal_type

    def __post_init__(self):
        """
        Post-initialization to ensure milestones don't have date ranges.
        Milestones are point-in-time checkpoints, not ranges.
        """
        # Clear start_date since milestones don't have ranges
        self.start_date = None

        # Ensure target_date is set for milestones
        if self.target_date is None:
            raise ValueError("Milestone requires a target_date")


@dataclass
class SmartGoal(Goal):
    """
    SMART goal with strict validation and required fields.

    Specific, Measurable, Achievable, Relevant, Time-bound

    Enforces that all SMART criteria are met:
    - Specific: Clear description
    - Measurable: Has unit and target value
    - Achievable: Target is positive
    - Relevant: Has relevance statement
    - Time-bound: Has start and end dates in future

    Examples:
    - "Run 120km in 10 weeks starting Oct 10"
    - "Read 40 hours of technical content by Dec 19"

    This is a grandchild of ThingIWant:
    ThingIWant → Goal → SmartGoal

    Note: Fields are inherited as Optional from Goal, but validation
    in __post_init__ ensures they are provided at construction time.
    """
    goal_type: str = 'SmartGoal'  # Override parent's goal_type

    def __post_init__(self):
        """Validate SMART criteria after initialization"""
        # Validate Measurable
        if not self.measurement_unit or not self.measurement_unit.strip():
            raise ValueError("SmartGoal requires measurement_unit (Measurable)")

        # Validate Achievable - need to check for None first since Optional[float]
        if self.measurement_target is None or self.measurement_target <= 0:
            raise ValueError(f"SmartGoal target must be positive, got {self.measurement_target} (Achievable)")

        # Validate Time-bound - need to check for None first
        if self.start_date is None or self.target_date is None:
            raise ValueError("SmartGoal requires both start_date and target_date (Time-bound)")
        if self.start_date >= self.target_date:
            raise ValueError("SmartGoal start_date must be before target_date (Time-bound)")

        # Validate Relevant
        if not self.how_goal_is_relevant or not self.how_goal_is_relevant.strip():
            raise ValueError("SmartGoal requires relevance statement (Relevant)")

        # Validate Achievable (actionable)
        if not self.how_goal_is_actionable or not self.how_goal_is_actionable.strip():
            raise ValueError("SmartGoal requires how_goal_is_actionable statement (Achievable)")

    @property
    def is_smart(self) -> bool:
        """Always True - this is a SMART goal by definition"""
        return True



# ===== KEY PATTERNS =====
#
# 1. INHERITANCE: Child classes GET all parent attributes automatically
#    - SmartGoal gets 'description' without defining it
#    - dataclasses handle super().__init__() automatically
#
# 2. SIBLINGS: Classes at same level differentiate by WHAT THEY ADD or HOW THEY BEHAVE
#    - Goal vs Aspiration vs Distraction all have description
#    - Different purposes/meanings in your domain
#
# 3. DATACLASS FIELDS: Instance attributes with automatic __init__ generation
#    - No need for manual __init__ methods
#    - Use field() for default_factory and other configurations
#    - __post_init__ for validation logic that runs after initialization
